usr/sbin/p11sak/p11sak.c

```diff
@@ -1652,146 +1783,231 @@ static CK_RV parse_gen_key_args(char *argv[], int argc, p11sak_kt *kt,
     }
 
     /* Check args */
-    ret = check_args_gen_key(kt, *keylength, *ECcurve);
+    rc = check_args_gen_key(kt, *keylength, *ECcurve);
 
     /* Check required options */
     if (*label == NULL) {
         printf("Key label must be specified.\n");
-        ret = CKR_ARGUMENTS_BAD;
+        rc = CKR_ARGUMENTS_BAD;
     }
 
     if (*slot == 0) {
         printf("Slot number must be specified.\n");
-        ret = CKR_ARGUMENTS_BAD;
+        rc = CKR_ARGUMENTS_BAD;
     }
 
-    return ret;
+    return rc;
 }
+/**
+ * Parse the remove-key args.
+ */
+static CK_RV parse_remove_key_args(char *argv[], int argc, p11sak_kt *kt,
+                                   CK_SLOT_ID *slot, char **pin, char **label,
+                                   CK_ULONG *keylength, CK_BBOOL *forceAll)
+{
+    CK_RV rc;
+    int i;
+
+    if (last_parm_is_help(argv, argc)) {
+        print_removekeys_help();
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    for (i = 2; i < argc; i++) {
+        /* Get arguments */
+        if (strcmp(argv[i], "DES") == 0 || strcmp(argv[i], "des") == 0) {
+            *kt = kt_DES;
+            *keylength = 64;
+        } else if (strcmp(argv[i], "3DES") == 0
+                || strcmp(argv[i], "3des") == 0) {
+            *kt = kt_3DES;
+        } else if (strcmp(argv[i], "AES") == 0 || strcmp(argv[i], "aes") == 0) {
+            *kt = kt_AES;
+        } else if (strcmp(argv[i], "RSA") == 0 || strcmp(argv[i], "rsa") == 0) {
+            *kt = kt_RSAPKCS;
+        } else if (strcmp(argv[i], "EC") == 0 || strcmp(argv[i], "ec") == 0) {
+            *kt = kt_EC;
+            /* Get options */
+        } else if (strcmp(argv[i], "--slot") == 0) {
+            *slot = (CK_ULONG) atol(argv[i + 1]);
+            i++;
+        } else if (strcmp(argv[i], "--pin") == 0) {
+            *pin = argv[i + 1];
```

Comment:
I think you have to check for an out of bounds read here: If "--poin" is the last string in argv, argv[i+1] becomes a NULL pointer deref.

Reply:
Included out of bounds check.